src/core/or/circuitpadding.c

```diff
@@ -0,0 +1,2283 @@
+
+/**
+ * Return the circpad_stat_t for the current state based on the
+ * mutable info.
+ */
+STATIC const circpad_state_t *
+circpad_machine_current_state(circpad_machineinfo_t *machine)
+{
+  switch (machine->current_state) {
+    case CIRCPAD_STATE_END:
+      return NULL;
+
+    case CIRCPAD_STATE_START:
+      return &CIRCPAD_GET_MACHINE(machine)->start;
+
+    case CIRCPAD_STATE_BURST:
+      return &CIRCPAD_GET_MACHINE(machine)->burst;
+
+    case CIRCPAD_STATE_GAP:
+      return &CIRCPAD_GET_MACHINE(machine)->gap;
+  }
+
+  log_fn(LOG_WARN,LD_CIRC,
+         "Invalid circuit padding state %d",
+         machine->current_state);
+  //tor_fragile_assert();
+
+  return NULL;
+}
+
+/**
+ * Calculate the lower bound of a histogram bin. The upper bound
+ * is obtained by calling this function with bin+1, and subtracting 1.
+ *
+ * The 0th bin has a special value -- it only represents start_usec.
+ * This is so we can specify a probability on 0-delay values.
+ *
+ * After bin 0, bins are exponentially spaced, so that each subsequent
+ * bin is twice as large as the previous. This is done so that higher
+ * time resolution is given to lower time values.
+ *
+ * The infinity bin is a the last bin in the array (histogram_len-1).
+ * It has a usec value of CIRCPAD_DELAY_INFINITE (UINT32_MAX).
+ */
+STATIC circpad_delay_t
+circpad_histogram_bin_to_usec(circpad_machineinfo_t *mi, int bin)
+{
+  const circpad_state_t *state = circpad_machine_current_state(mi);
+  circpad_delay_t start_usec;
+
+  if (state->use_rtt_estimate)
+    start_usec = mi->rtt_estimate_us+state->start_usec;
+  else
+    start_usec = state->start_usec;
+
+  if (bin >= state->histogram_len-1)
+    return CIRCPAD_DELAY_INFINITE;
+
+  if (bin == 0)
+    return start_usec;
+
+  if (bin == 1)
+    return start_usec+1;
+
+  return start_usec
+      + (state->range_sec*USEC_PER_SEC)/(1<<(state->histogram_len-bin+1));
+}
+
+/**
+ * Return the bin that contains the usec argument.
+ * "Contains" is defined as us in [lower, upper).
+ *
+ * This function will never return the infinity bin (histogram_len-1),
+ * in order to simplify the rest of the code.
+ *
+ * This means that technically the last bin (histogram_len-2)
+ * has range [start_usec+range_sec*USEC_PER_SEC,
+ * CIRCPAD_DELAY_INFINITE].
+ */
+STATIC int
+circpad_histogram_usec_to_bin(circpad_machineinfo_t *mi, circpad_delay_t us)
+{
+  const circpad_state_t *state = circpad_machine_current_state(mi);
+  circpad_delay_t start_usec;
+  int bin;
+
+  /* Our state should have been checked to be non-null by the caller
+   * (circpad_machine_remove_token()) */
+  if (BUG(state == NULL)) {
+    return 0;
+  }
+
+  if (state->use_rtt_estimate)
+    start_usec = mi->rtt_estimate_us+state->start_usec;
+  else
+    start_usec = state->start_usec;
+
+  if (us <= start_usec)
+    return 0;
+
+  if (us == start_usec+1)
+    return 1;
+
+  bin = state->histogram_len -
+    tor_log2((state->range_sec*USEC_PER_SEC)/(us-start_usec+1));
```

Comment:
This line contains 6 values and 6 operations, please split for maintainability.

Reply:
5fa69f86ae6c4843599e2ad994fffb814e26aae4